factory-containers/compose_app_downloader.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/python3
+
+import yaml
+import subprocess
+import logging
+import time
+import os
+import tempfile
+import shutil
+import argparse
+import contextlib
+
+
+logger = logging.getLogger(__name__)
+
+
+class ComposeApp:
+    def __init__(self, name, file_path):
+        self.name = name
+        self.file_path = file_path
+
+        with open(self.file_path) as compose_file:
+            self._composed_app = yaml.safe_load(compose_file)
+
+    def services(self):
+        return self._composed_app['services'].items()
+
+
+class ContainerDaemon:
+    def __init__(self):
+        self.cmd = 'containerd'
+
+    def __enter__(self):
+        self._container_dir = tempfile.mkdtemp()
+
+        self._root_dir = os.path.join(self._container_dir, 'root')
+        os.mkdir(self._root_dir)
+
+        self._state_dir = os.path.join(self._container_dir, 'state')
+        os.mkdir(self._state_dir)
+
+        self.address = os.path.join(self._container_dir, 'containerd.sock')
+
+        logger.info("Starting containerd for docker daemon...")
+        self._process = subprocess.Popen([self.cmd, '--address', self.address,
+                                          '--root', self._root_dir,
+                                          '--state', self._state_dir],
+                                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # hack, let containerd some time to start
+        time.sleep(2)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._process.terminate()
+        self._process.wait(timeout=60)
+
+        logger.info("Containerd daemon has been stopped")
+        shutil.rmtree(self._container_dir)
+
+
+class DockerDaemon:
```

Comment:
we already have dockerd running when this code gets hit. Do you need your own instance to make sure you have a pristine environment?

Reply:
Yes, there are a few reasons for that (the one you mentioned is valid too):
1. a host environment executing the given may have or may not have dockerd running. The given approach makes the implementation less vulnerable/sensitive to a host env, so it can work on a local host, VM or any other container as long as dockerd is installed;

2. The host dockerd's data root (/var/lib/docker/) can be polluted by data of some other images while we need data/files that belongs solely to the target app images;

3. Using `--platform` option requires running dockerd in the experimental mode. `"--platform" is only supported on a Docker daemon with experimental features enabled`. This option is required because we want to dump images of other than a host architecture (dumping arm64's images on amd64 host).

4. It allows dumpling images of all target architectures concurrently, multiple instances of DockerDaemon can be running concurrently without interfering each other what can speed up an overall process. Not implemented yet though :)